src/cdomains/intDomain.ml

```diff
@@ -2282,7 +2293,28 @@ struct
   let ge ik x y = le ik y x
 
   let bitnot = lift1 BigInt.bitnot
-  let bitand = lift2 BigInt.bitand
+
+  let bitand ik x y = norm ik (match x,y with
+      (* We don't bother with exclusion sets: *)
+      | `Excluded _, `Definite i ->
+        (* Except in two special cases *)
+        if BigInt.equal i BigInt.zero then
+          `Definite BigInt.zero
+        else if BigInt.equal i BigInt.one then
+          of_interval IBool (BigInt.zero, BigInt.one)
```

Comment:
This is giving the `bitand` result `IBool` ikind which doesn't seem right. I think it should still be the original ikind, just the range is smaller. That's how you've implemented it for intervals as well.

Reply:
I don't think it's problematic, inside this domain the `ikind` is not stored and only used to compute the ranges:

https://github.com/goblint/analyzer/blob/4d18300af5f6ae1b04bdbf29b29775af07fbc96c/src/cdomains/intDomain.ml#L2157

https://github.com/goblint/analyzer/blob/4d18300af5f6ae1b04bdbf29b29775af07fbc96c/src/cdomains/intDomain.ml#L1989

Reply:
Hmm, that makes it quite misleading though. The result is based on `IBool`, not `(BigInt.zero, BigInt.one)` at all. So right now this is the only way anyway.

But we should open an issue about fixing `DefExc.of_interval` to actually derive a bit range from the actual given interval and intersect it with the ikind's range. I didn't check but `starting` and `ending` should work similarly. It might make a difference because base analysis itself uses `of_interval`/`starting`/`ending`.

Reply:
I guess I'll leave the code as it currently is in light of #1159.